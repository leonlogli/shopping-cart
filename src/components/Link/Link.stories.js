import React from 'react'
import Link from './Link'

export default {
  title: 'Link',
}

export const RoutingLink = () => <Link href="/home">Routing Link</Link>

export const NonRoutingLink = () => (
  <Link to="http//test.com">Non Routing Link</Link>
)
