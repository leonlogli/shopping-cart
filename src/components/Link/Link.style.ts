import styled from 'styled-components'
import Link from '@material-ui/core/Link'

const LinkBase = styled(Link)`
  cursor: pointer;
`

export { LinkBase }
export default LinkBase
