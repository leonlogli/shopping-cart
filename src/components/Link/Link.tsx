import React, { forwardRef } from 'react'
import NextLink, { LinkProps as NextLinkProps } from 'next/link'
import { LinkProps as MuiLinkProps } from '@material-ui/core/Link'

import LinkBase from './Link.style'

export type LinkProps = Omit<MuiLinkProps, 'href'> &
  Pick<NextLinkProps, 'as'> & {
    href?: NextLinkProps['href']
    /**
     * Href for non routing link. Do not use it with href
     */
    to?: string
    /**
     * The component used for the root node. Either a string to use a HTML element or a component.
     */
    component?: keyof JSX.IntrinsicElements | React.ComponentType<any>
  }

/**
 * Link is a component which handle routing and non-routing href
 */
const Link = forwardRef<HTMLLinkElement, LinkProps>((props, ref) => {
  const { as, href, to, ...other } = props

  if (!href) {
    return <LinkBase ref={ref} href={to} {...other} />
  }

  return (
    <NextLink href={href as any} as={as} passHref>
      <LinkBase ref={ref} {...other} />
    </NextLink>
  )
})

export { Link }
export default Link
