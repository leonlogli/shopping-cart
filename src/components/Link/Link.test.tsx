import React from 'react'
import { mount } from 'enzyme'

import Link from './Link'

describe('<Link />', () => {
  const wrapper = mount(<Link href="/about">Link</Link>)

  it('should render children', () => {
    expect(wrapper.find('a').length).toEqual(1)
  })

  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should behave like non rounting', () => {
    const wrapper2 = mount(<Link to="http//test.com">Link</Link>)

    expect(wrapper2.find('a').props().href).toBe('http//test.com')
  })

  it('should render button', () => {
    // eslint-disable-next-line jsx-a11y/anchor-is-valid
    const wrapper2 = mount(<Link component="button">Link</Link>)

    expect(wrapper2.find('button').length).toBe(1)
  })
})
