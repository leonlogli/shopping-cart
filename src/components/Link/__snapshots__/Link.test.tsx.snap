// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should match a snapshot 1`] = `
<ForwardRef
  href="/about"
>
  <Link
    href="/about"
    passHref={true}
  >
    <Styled(WithStyles(ForwardRef(Link)))
      href="/about"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <WithStyles(ForwardRef(Link))
        className="jxeIVE"
        href="/about"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <ForwardRef(Link)
          className="jxeIVE"
          classes={
            Object {
              "button": "MuiLink-button",
              "focusVisible": "Mui-focusVisible",
              "root": "MuiLink-root",
              "underlineAlways": "MuiLink-underlineAlways",
              "underlineHover": "MuiLink-underlineHover",
              "underlineNone": "MuiLink-underlineNone",
            }
          }
          href="/about"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <WithStyles(ForwardRef(Typography))
            className="MuiLink-root MuiLink-underlineHover jxeIVE"
            color="primary"
            component="a"
            href="/about"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            variant="inherit"
          >
            <ForwardRef(Typography)
              className="MuiLink-root MuiLink-underlineHover jxeIVE"
              classes={
                Object {
                  "alignCenter": "MuiTypography-alignCenter",
                  "alignJustify": "MuiTypography-alignJustify",
                  "alignLeft": "MuiTypography-alignLeft",
                  "alignRight": "MuiTypography-alignRight",
                  "body1": "MuiTypography-body1",
                  "body2": "MuiTypography-body2",
                  "button": "MuiTypography-button",
                  "caption": "MuiTypography-caption",
                  "colorError": "MuiTypography-colorError",
                  "colorInherit": "MuiTypography-colorInherit",
                  "colorPrimary": "MuiTypography-colorPrimary",
                  "colorSecondary": "MuiTypography-colorSecondary",
                  "colorTextPrimary": "MuiTypography-colorTextPrimary",
                  "colorTextSecondary": "MuiTypography-colorTextSecondary",
                  "displayBlock": "MuiTypography-displayBlock",
                  "displayInline": "MuiTypography-displayInline",
                  "gutterBottom": "MuiTypography-gutterBottom",
                  "h1": "MuiTypography-h1",
                  "h2": "MuiTypography-h2",
                  "h3": "MuiTypography-h3",
                  "h4": "MuiTypography-h4",
                  "h5": "MuiTypography-h5",
                  "h6": "MuiTypography-h6",
                  "noWrap": "MuiTypography-noWrap",
                  "overline": "MuiTypography-overline",
                  "paragraph": "MuiTypography-paragraph",
                  "root": "MuiTypography-root",
                  "srOnly": "MuiTypography-srOnly",
                  "subtitle1": "MuiTypography-subtitle1",
                  "subtitle2": "MuiTypography-subtitle2",
                }
              }
              color="primary"
              component="a"
              href="/about"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              variant="inherit"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover jxeIVE MuiTypography-colorPrimary"
                href="/about"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
              >
                Link
              </a>
            </ForwardRef(Typography)>
          </WithStyles(ForwardRef(Typography))>
        </ForwardRef(Link)>
      </WithStyles(ForwardRef(Link))>
    </Styled(WithStyles(ForwardRef(Link)))>
  </Link>
</ForwardRef>
`;
